src/tap/src/tapcell.cpp

```diff
@@ -57,17 +63,1091 @@ Tapcell::~Tapcell()
+  Tapcell::filled_sites.clear();
+  if (endcap_master != nullptr ) {
+    insertEndcaps(rows, endcap_master, cnrcap_masters, endcap_prefix);
+  }
+  if (add_boundary_cell) {
+    std::vector<std::string> tap_nw_masters;
+    tap_nw_masters.push_back(tap_nwintie_master);
+    tap_nw_masters.push_back(tap_nwin2_master);
+    tap_nw_masters.push_back(tap_nwin3_master);
+    tap_nw_masters.push_back(tap_nwouttie_master);
+    tap_nw_masters.push_back(tap_nwout2_master);
+    tap_nw_masters.push_back(tap_nwout3_master);
+
+    std::vector<std::string> tap_macro_masters;
+    tap_macro_masters.push_back(incnrcap_nwin_master);
+    tap_macro_masters.push_back(tap_nwin2_master);
+    tap_macro_masters.push_back(tap_nwin3_master);
+    tap_macro_masters.push_back(tap_nwintie_master);
+    tap_macro_masters.push_back(incnrcap_nwout_master);
+    tap_macro_masters.push_back(tap_nwout2_master);
+    tap_macro_masters.push_back(tap_nwout3_master);
+    tap_macro_masters.push_back(tap_nwouttie_master);
+
+    insertAtTopBottom(rows, tap_nw_masters, db_->findMaster(cnrcap_nwin_master), endcap_prefix);
+    insertAroundMacros(rows, tap_macro_masters, db_->findMaster(cnrcap_nwin_master), endcap_prefix);
+  }
+  insertTapcells(rows, tapcell_master, dist, tap_prefix);
+  Tapcell::filled_sites.clear();
+}
+
+void Tapcell::cutRows(odb::dbMaster* endcap_master, std::vector<odb::dbBox*> blockages,int& halo_x, int& halo_y)
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  int rows_count = (block->getRows()).size();
+  int block_count = blockages.size();
+  int min_row_width;
+  int cut_rows_count;
+  int end_width;
+
+  if (endcap_master != nullptr) {
+    end_width = endcap_master->getWidth();
+    min_row_width = 2*end_width;
+  } else {
+    min_row_width = 0;
+  }
+  
+  //Gather rows needing to be cut up front
+  std::vector<odb::dbRow*> blocked_rows;   
+  std::map<std::string, std::vector<odb::dbBox*>> row_blockages;
+  std::string row_name;
+  for (odb::dbBox* blockage : blockages) {
+    for(odb::dbRow* row : block->getRows()) {
+      if (overlaps(blockage, row, halo_x, halo_y)) {
+        row_name = row->getName();
+        if (row_blockages.find(row_name) == row_blockages.end()) {
+          blocked_rows.push_back(row);
+          std::vector<odb::dbBox*> row_blockages_bbox;
+          row_blockages_bbox.push_back(blockage);
+          row_blockages.insert(std::pair<std::string, vector<odb::dbBox*>>(row_name, row_blockages_bbox));
+        } else {
+          row_blockages[row_name].push_back(blockage); 
+        }
+      }
+    }
+  }
+  
+  // cut rows around macros
+  for(odb::dbRow* row : blocked_rows) {
+    cutRow(block, row, row_blockages, min_row_width, halo_x, halo_y);
+  }
+  cut_rows_count = ((block->getRows()).size())-rows_count;
+  logger_->info(utl::TAP, 1, "Macro blocks found: {}", block_count);
+  logger_->info(utl::TAP, 2, "Original rows: {}", rows_count);
+  logger_->info(utl::TAP, 3, "Created rows: {}", cut_rows_count);
```

Comment:
"Found {} macro blocks."
"Created {} rows for a total of {} rows."
Is total/final number of rows more useful than original?